pmdarima/arima/arima.py

```diff
@@ -1196,8 +1196,9 @@ def plot_diagnostics(self, variable=0, lags=10, fig=None, figsize=None):
         ax = fig.add_subplot(222)
         # temporarily disable Deprecation warning, normed -> density
         # hist needs to use `density` in future when minimum matplotlib has it
+        # normed keyword arugment is no longer supported in matplotlib since version 3.2.0
         with warnings.catch_warnings(record=True):
-            ax.hist(resid_nonmissing, normed=True, label='Hist')
+            ax.hist(resid_nonmissing, density=True, label='Hist')
```

Comment:
This will break backwards compatibility with older versions. Can we we add some sort of `ax_hist` compatibility function in [`pmdarima.compat.matplotlib`](https://github.com/alkaline-ml/pmdarima/blob/master/pmdarima/compat/matplotlib.py) that uses `density` in versions > 3.2.0 and `normed` in previous versions?

Reply:
The `density` keyword is present in matplotlib since version 2.1.0 (released on October 7 2017): https://matplotlib.org/2.1.0/api/_as_gen/matplotlib.pyplot.hist.html
It should be compatible starting form that version.

I looked at the source code of statsmodels that is called by `plot_diagnostics` to verify whether there was a requirement of the matplotlib version, but there seems to be none: it simply tries to import matplotlib, otherwise it raises an error.

Thus, you are right that this would break compatibility with previous matplotlib versions (or that a requirement would need to be added in pmdarima for `matplotlib>=2.1.0`, which is not optimal).

To resolve this issue, I took your idea and created the function `mpl_hist_arg` in `pmdarima.compat.matplotlib`, which returns `True` if `matplotlib>=2.1.0` and otherwise returns `False`.
It also returns `True` if the matplotlib version cannot be evaluated.
If `mpl_hist_arg=True`, `ax.hist` uses `density=True`, otherwise `normed=True` is used.

Since I'm relatively new to Python, there might be a more elegant/Pythonic way to acheive this, instead of repeating twice the `ax.hist` call in the code?

Reply:
Your new solution is certainly functional, but since you say you're new to Python, I hope you'll accept some constructive feedback towards a more "Pythonic" suggestion.

`normed` and `density` are what are known as [`kwargs`](https://stackoverflow.com/a/1769475/10696164) (Key-Word arguments), meaning they don't need to be supplied, but we _can_ supply them as a key-value pair. The trick is, we need to dynamically either supply `normed=True` or `density=True`. The way to do this is to supply a dictionary with the key-value pair we want, and ["unpack" it](https://www.python.org/dev/peps/pep-0448/) using `**`.

So, with that being said, I think we could re-write your compat function something like this:

```python
def mpl_hist_arg(value=True):
    """Find the appropriate `density` kwarg for our given matplotlib version.

    This will determine if we should use `normed` or `density`. Additionally,
    since this is a kwarg, the user can supply a value (True or False) that
    they would like in the output dictionary.

    Parameters
    ----------
    value : bool, optional (default=True)
        The boolean value of density/normed

    Returns
    -------
    density_kwarg : dict
        A dictionary containing the appropriate density kwarg for the
        installed  matplotlib version, mapped to the provided or default
        value
    """
    import matplotlib
    
    density_kwarg = 'density' if matplotlib.__version__ >= '2.1.0' else 'normed'
    return {density_kwarg: value}
```

Then, we can cleanly implement it like this:

```python
ax.hist(resid_nonmissing, label='Hist', **mpl_compat.mpl_hist_arg())
```

Since they are key word arguments, it doesn't matter that we changed the order. But key word arguments should generally be last as to not interfere with positional arguments. Additionally, we already have a check in `plot_diagnostics` to make sure `matplotlib` is installed, so that is why I left it out of my re-written function.

Let me know if that makes sense, but I think that would be a clean way to implement this without having to duplicate code

Reply:
Thank you for the feedback, it is quite appreciated!
I am indeed quite new to Python, I just started learning last month.
I had read about `kwargs` before, but your explanation really made me understand the concept.

I will edit the code as per your example.

Should we keep catching the deprecation warning for backwards compatibility (in case the `normed=True` is used)?
`with warnings.catch_warnings(record=True):`

Reply:
Awesome, looks good now!

Yeah, I think catching the warnings is still good. For anyone that has an older version of `matplotlib` installed, we don't want to suddenly start raising warnings in their workflows.